Test unit testing (#197)
SW709E19/Report: report/ChapterTest/UnitTest.tex

\section{Unit Tests}\label{sec:UnitTest}

> This sections describes the unit tests of the isolated functionality in the system.

Reviewer: ```suggestion
This section describes the unit tests of the isolated functionality in the system.
```

> This sections describes the unit tests of the isolated functionality in the system.

Reviewer 2: ```suggestion
This section describes the unit tests of the isolated functionality in the system.
```

> This sections describes the unit tests of the isolated functionality in the system.

Reviewer 3: ```suggestion
This section describes the unit tests of the functions in Google Cloud.
```

> Most of the components in the systems are used to integrate different components.

Reviewer 4: Not sure this makes sense

> Most of the components in the systems are used to integrate different components.

Reviewer: Den her sætning er ikke særlig god.
```suggestion
Most of the components in the system are used to integrate different components.
```

> Unit testing these components are difficult because it would require extensive mocking of dependencies in order to create an isolated unit for testing.

Reviewer: ```suggestion
Unit testing these components is difficult because it would require extensive mocking of dependencies in order to create an isolated unit for testing.
```

> Unit testing these components are difficult because it would require extensive mocking of dependencies in order to create an isolated unit for testing.

Reviewer 2: ```suggestion
Unit testing these components are difficult because it would require extensive mocking of dependencies to create an isolated unit for testing.
```

> Unit testing these components are difficult because it would require extensive mocking of dependencies in order to create an isolated unit for testing.

Reviewer 3: ```suggestion
Unit testing these components are difficult because it would require extensive mocking of dependencies in order to isolate components for testing.
```

It is also difficult to test sensor device firmware separately because it depends on specific hardware.

> Therefore, the unit tests has been limited to test the important backend functionality used to decode and map samples.

Reviewer: ```suggestion
Therefore, the unit tests have been limited to test the important backend functionality used to decode and map samples.
```

> Therefore, the unit tests has been limited to test the important backend functionality used to decode and map samples.

Reviewer 2: ```suggestion
Therefore, the unit tests have been limited to test the important backend functionality used to decode and map samples.
```

\subsection{Decoding}\label{sec:Decoding}

> Sensor devices send samples in the byte encoding shown in appendix X. 

Reviewer 4: Appendix X

This was done to reduce the volume of data transferred from sensor devices to the backend. 
This makes the decoding functionality in the Store Sample function a critical part of the system since an error can cause wrong values. 

> To prevent this unit tests has been developed to test this functionality. 

Reviewer: ```suggestion
To prevent this unit tests have been developed to test this functionality. 
```

\autoref{lst:decoding} shows a unit test of the decoding function.

\begin{lstlisting}[label={lst:decoding}, caption={Unit test of function decoding sample from byte encoding.}, captionpos=b]
var gateway1Encoding = Buffer.from([103, 97, 116, 101, ...]);

var expectedGateway1Decoding = { "version": 1, "gateway": { "encoding": "gateway-1", "version": "1", "id": "test-gateway", ... }, ... };

describe('sample-decoding v1', function () {
  describe('decode', function () {
    it('gateway-1 encoding', function () {
      assert.deepEqual(decode.decode(gateway1Encoding), expectedGateway1Decoding);
    });
    ...
  });
});
\end{lstlisting}

The test in \autoref{lst:decoding} calls the decoding function with a predefined byte encoding of a sample and compares it with the expected value.

\subsection{Mapping Samples to Time Series}
The Get Samples function combines measurements from samples in a given period into time series data. 
It is important that these measurements are merged correctly. 
Measurements in samples do not have a global timestamp but instead a local time in microseconds since the device started sampling. 

> When merging samples the time for each measurement is estimated based on the global timestamp of when the sample was received. 

Reviewer: ```suggestion
When merging samples, the time for each measurement is estimated based on the global timestamp of when the sample was received. 
```
